libs/meta/include/meta/tuple.hpp

```diff
@@ -17,38 +17,79 @@
 //
 //------------------------------------------------------------------------------
 
+#include "meta/param_pack.hpp"
+
 #include <cstddef>
 #include <tuple>
 #include <utility>
 
 namespace fetch {
 namespace meta {
 
-template <typename T, typename... Ts>
+template <std::size_t, typename...>
+struct SliceTuple;
+template <std::size_t N, typename... Ts>
+struct SliceTuple<N, std::tuple<Ts...>>
+{
+private:
+  static_assert(N <= sizeof...(Ts), "Slice index exceeds tuple size");
+
+  static constexpr std::size_t start_length = N;
+  static constexpr std::size_t end_length   = sizeof...(Ts) - N;
+
+  template <std::size_t... Is>
+  static std::tuple<std::tuple_element_t<Is, std::tuple<Ts...>>...> get_start(
+      std::index_sequence<Is...>);
+  template <std::size_t... Is>
+  static std::tuple<std::tuple_element_t<N + Is, std::tuple<Ts...>>...> get_end(
+      std::index_sequence<Is...>);
+
+public:
+  using Head = decltype(get_start(std::make_index_sequence<start_length>()));
+  using Tail = decltype(get_end(std::make_index_sequence<end_length>()));
+
+  static_assert(std::tuple_size<Head>::value == start_length, "");
+  static_assert(std::tuple_size<Tail>::value == end_length, "");
+};
+
+namespace internal {
+
+template <typename... Ts>
 struct RemoveFirstType;
-template <typename T, typename... Ts>
-struct RemoveFirstType<std::tuple<T, Ts...>>
+template <typename... Ts>
+struct RemoveFirstType<std::tuple<Ts...>>
 {
-  using type = typename std::tuple<Ts...>;
+  using type = typename SliceTuple<1, std::tuple<Ts...>>::Tail;
 };
 
 template <typename... Ts>
 struct RemoveLastType;
 template <typename... Ts>
 struct RemoveLastType<std::tuple<Ts...>>
 {
-  template <std::size_t... Is>
-  static std::tuple<std::tuple_element_t<Is, std::tuple<Ts...>>...> f(std::index_sequence<Is...>);
-  using type = decltype(f(std::make_index_sequence<sizeof...(Ts) - 1u>()));
+  using type = typename SliceTuple<sizeof...(Ts) - 1, std::tuple<Ts...>>::Head;
 };
 
 template <typename... Ts>
 struct GetLastType;
 template <typename... Ts>
 struct GetLastType<std::tuple<Ts...>>
 {
-  using type = typename std::tuple_element_t<sizeof...(Ts) - 1u, std::tuple<Ts...>>;
+  using type =
+      std::tuple_element_t<0, typename SliceTuple<sizeof...(Ts) - 1, std::tuple<Ts...>>::Tail>;
```

Comment:
You're not wrong, but given that there is no runtime overhead (`make_index_sequence` is only called in an unevaluated context here), I'm prepared to live with it and prefer keeping the source code consistent. If we ever deal with tuples thousands of elements long, it might make sense to worry about meta-algorithm complexity like this. Right now our build times face far worse bottlenecks in the form of poor modularisation and bloated headers.